deployer/helm_upgrade_decision.py

```diff
@@ -0,0 +1,367 @@
+test_env = os.getenv("RUN_ENV", False)
+
+
+def discover_modified_common_files(modified_paths: list):
+    """There are certain common files which, if modified, we should upgrade all hubs
+    and/or all clusters appropriately. These common files include the helm charts we
+    deploy, as well as the GitHub Actions and deployer package we use to deploy them.
+    Args:
+        modified_paths (list[str]): The list of files that have been added or modified
+            in a given GitHub Pull Request.
+    Returns:
+        upgrade_all_clusters (bool): Whether or not all clusters should be upgraded
+            since the support chart has changed
+        upgrade_all_hubs (bool): Whether or not all hubs on all clusters should be
+            upgraded since a core piece of infrastructure has changed
+    """
+    # If any of the following filepaths have changed, we should update all hubs on all clusters
+    common_filepaths = [
+        "deployer/*",
+        "helm-charts/basehub/*",
+        "helm-charts/daskhub/*",
+        ".github/actions/deploy/*",
+        "requirements.txt",
+        ".github/workflows/deploy-hubs.yaml",
+    ]
+    # If this filepath has changes, we should update the support chart on all clusters
+    support_chart_filepath = "helm-charts/support/*"
+
+    # Discover if the support chart has been modified
+    upgrade_all_clusters = bool(fnmatch.filter(modified_paths, support_chart_filepath))
+
+    # Discover if any common config has been modified
+    upgrade_all_hubs = False
+    while not upgrade_all_hubs:
+        upgrade_all_hubs = bool(
+            fnmatch.filter(modified_paths, common_filepath_pattern)
+            for common_filepath_pattern in common_filepaths
+        )
+
+    return upgrade_all_clusters, upgrade_all_hubs
+
+
+def get_categorised_unique_filepaths(added_or_modified_files: list):
+    """For a list of added and modified files, get the following categorised, unique
+    filepaths:
+    - A list of unique filepaths to cluster folders containing added/modified files
+    - A set of all added/modified files matching the pattern "*/cluster.yaml"
+    - A set of all added/modified files matching the pattern "*/*.values.yaml"
+    - A set of all added/modified files matching the pattern "*/support.values.yaml"
+
+    Note: "cluster folders" are those that contain a cluster.yaml file.
+
+    Args:
+        added_or_modified_files (list[str]): A list of files that have been added or
+            modified in a GitHub Pull Request
+
+    Returns:
+        list[path obj]: A list of unique filepaths to cluster folders
+        set[str]: A set of all files matching the pattern "*/cluster.yaml"
+        set[str]: A set of all files matching the pattern "*/*.values.yaml"
+        set[str]: A set of all files matching the pattern "*/*support*.values.yaml"
+    """
+    patterns_to_match = ["*/cluster.yaml", "*/*.values.yaml", "*/*support*.values.yaml"]
```

Comment:
_Original discussion:_ https://github.com/2i2c-org/infrastructure/pull/1093#discussion_r826196267

> With this, we will start relying on (but not enforcing here) the previously non-strict naming convention of having values.yaml in the end of all values files.
> 
> Adding this strict filter pattern makes me think we should enforce `values.yaml` suffix in `helm_chart_values_files`.
> 
> I'm brainstorming an idea that one could not need a filter pattern at all by detecting/inspecting all clusters and inspecting their `helm_chart_values_yaml` files etc for relevant files to observe for changes. Given how this implementation is now though, I suggest its pair with enforcement of the files listed in `helm_chart_values_files` will lead to matching the patterns defined here.

Reply:
@consideRatio I think I avoided "reading all cluster.yaml files first" as it felt a bit too invasive. Instead, I used the filtering to find unique cluster folders that had additions/modifications and just read those.

Reply:
I figure two strategies are possible and plausible:
a) to rely on assumptions about files' paths and names
b) to rely on explicit configuration

I'm leaning towards thinking b) could be better with ideas like:
- Letting a computer read all cluster.yaml files seems like a suitable chore for a computer
- Looking for specific filename patterns adds logic we didn't really need as we had explicit configuration to identify the relevant files
- Looking for relevant files with new logic based on conventions seems to force us to following them while we previously just recommended them, but forcing isn't enforcing, so now we also need to enforce they are followed so our logic doesn't break.
  - File names
  - Where files are located
     For example I think we could use a file shared by multiple clusters even it wasn't recommended, but now I think all files must be located under the cluster associated directly

Reply:
My most recent work removes the majority of the file filtering I was doing and instead does the following:

- From the input list of added/modified files, get unique cluster filepaths and verify these are cluster filepaths by checking for the existence of a cluster.yaml file
- For each identified unique cluster filepath, read in the cluster.yaml file and get the list of helm chart values files (whether for a hub or support)
- Inspect whether any of these files pulled from cluster.yaml appear in the original list of added/modified files parsed from the command line and decide whether that hub needs upgrading or support for that cluster needs upgrading based on the results of this inspection

We only read in _all_ cluster.yaml files if core infrastructure has changed, random e.g., the daskhub helm chart